### PR TITLE
Gan pantry

### PR DESCRIPTION
**Pull request by ArianaFreitag**
Setting up gan-pantry

### CODE REVIEW

**Comment by ccurro** [comment]
Note: We will want a script to run classifier tests separate from full gan tests

File: src/train.py [hunk 1]
```diff
@@ -101,14 +123,28 @@ def main(argv):
...
+            style_vector = torch.rand(500, device=DEVICE)
+            char_vector = charVector()
+
+            #the generator is provided by the pantry.py
+            charGenerated = generator.generate(char_vector, style_vector)
 
+            # Zero the parameter gradients #weird investigate further
+            optimizerDisc.zero_grad()
+
+            # Train the optimizer
```
Comment (eigenfoo) [inline src/train.py]: We don't train the optimizer! The optimizer is needed to train the network :innocent:

Comment (ArianaFreitag) [reply src/train.py]: i think typo hahaha

File: src/train.py [hunk 2]
```diff
@@ -83,16 +88,33 @@ def validate(net, epoch_num):
     print('\n\n\n')
 
 
+def charVector(batch,char=None):
+    # todo fix tensor
+    if not char:
+        char = np.random.randint(0, 70, batch)
+    
+    onehotchar = np.zeros(70)
+    onehotchar[char] = 1
```
Comment (eigenfoo) [inline src/train.py]: Be careful: we've been bitten in the past by incorrect datatypes. This looks like its `np.int64`. Do you intend for it to be `np.float64` instead?

Comment (ArianaFreitag) [reply src/train.py]: i think we want it to be an int, bc int corresponds with the char index

File: src/resnet.py [hunk 3]
```diff
@@ -73,6 +73,8 @@ def __init__(self,
         self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
         self.fc = nn.Linear(256 * block.expansion, num_classes)
 
+        self.sigmoid = nn.Sigmoid()
+
```
Comment (ccurro) [inline src/resnet.py]: There's 2 changes in `src/resnet.py` are either necessary for this PR?

File: src/train.py [hunk 4]
```diff
@@ -8,17 +8,23 @@
...
+    "Pastry: indicates discriminator network, optimizer and number of epochs")
 flags.mark_flag_as_required("pastry")
+flags.DEFINE_string("bread", None, "Bread: indicates generator and optimizer ")
+flags.mark_flag_as_required("bread")
 
 DATA_PATH = '/flour/noCapsnoRepeatsSingleExampleProtos/'
+# DATA_PATH  = '/home/ostap/Documents/fontbakers/data/train/'
 DIMENSIONS = (20, 30, 3, 2)
 TRAIN_TEST_SPLIT = 0.8
+BATCH_SIZE = 16
```
Comment (ccurro) [inline src/train.py]: Make this a flag

File: src/train.py [hunk 5]
```diff
@@ -101,18 +135,37 @@ def main(argv):
...
+
+            # Generate characters
+            charsGenerated = gen(char_vec, style_vector)
+
+            # Zero the parameter gradients for the discriminator
+            optimizerDisc.zero_grad()
+
+            # Train the discriminator
+            prediction = disc(font)
+            lossDisc = criterionDisc(prediction, labels)
```
Comment (ccurro) [inline src/train.py]: The discriminator should be updated using the generated as well, not just the real data.

File: src/resnet.py (hunk 3 above)
Comment (ArianaFreitag) [reply src/resnet.py]: no

File: src/train.py (hunk 4 above)
Comment (ArianaFreitag) [reply src/train.py]: batch size?

Comment (ccurro) [reply src/train.py]: Yes

File: src/train.py [hunk 6]
```diff
@@ -8,17 +8,25 @@
...
 import pantry
+from fridge import model_saver as saver
 
 np.set_printoptions(threshold=np.inf)  # Print full confusion matrix.
 FLAGS = flags.FLAGS
+
 flags.DEFINE_string(
     "pastry", None,
-    "Pastry: indicates network, optimizer and number of epochs")
+    "Pastry: indicates discriminator network, optimizer and number of epochs")
```
Comment (eigenfoo) [inline src/train.py]: Eventually we should move the number of epochs out as a separate flag. I'll do that in a subsequent PR.

File: src/train.py [hunk 7]
```diff
@@ -28,6 +36,7 @@
     'five', 'six', 'seven', 'eight', 'nine', 'exclam', 'numbersign', 'dollar',
     'percent', 'ampersand', 'asterisk', 'question', 'at'
 ]
+
```
Comment (eigenfoo) [inline src/train.py]: Removing the newline... 
```suggestion
```

File: src/train.py [hunk 8]
```diff
@@ -17,14 +17,16 @@
     "pastry", None,
     "Pastry: indicates discriminator network, optimizer and number of epochs")
 flags.mark_flag_as_required("pastry")
-flags.DEFINE_string("bread", None, "Bread: indicates generator and optimizer ")
+flags.DEFINE_string("bread", None, "Bread: indicates generator and optimizer")
 flags.mark_flag_as_required("bread")
+flags.DEFINE_string("batch", None, "Batch: indicates batch size")
```
Comment (ccurro) [inline src/train.py]: Wrong type for the flag. `flags.DEFINE_integer`

File: src/train.py [hunk 9]
```diff
@@ -144,18 +146,21 @@ def main(argv):
...
 
+            # Make predictions
+            predictionReal = disc(font)
+            predictionFake = disc(charsGenerated)
+            predictionCombined = torch.cat((predictionReal, predictionFake), 0)
+
             # Train the discriminator
-            prediction = disc(font)
-            lossDisc = criterionDisc(prediction, labels)
+            lossDisc = criterionDisc(predictionCombined, labels)
```
Comment (ccurro) [inline src/train.py]: You also need to do a concat for labels. Right now labels just contains the real.

File: src/train.py [hunk 10]
```diff
@@ -101,18 +137,44 @@ def main(argv):
...
+            optimizer_disc.zero_grad()
+
+            # Make predictions
+            prediction_real = disc(font)
+            prediction_fake = disc(chars_generated)
+            prediction_combined = torch.cat((prediction_real, prediction_fake),
+                                            0)
+
+            # combine char vector from real and fake CHARACTERS
+            labels_combined = torch.cat((labels, char_vec), 0)
```
Comment (ccurro) [inline src/train.py]: This should not be with `char_vec`. The discriminator needs to know that the fake is fake, not what they are intended to be.

File: src/train.py [hunk 11]
```diff
@@ -8,17 +8,24 @@
...
     "pastry", None,
-    "Pastry: indicates network, optimizer and number of epochs")
+    "Pastry: indicates discriminator network, optimizer and number of epochs")
 flags.mark_flag_as_required("pastry")
+flags.DEFINE_string("bread", None, "Bread: indicates generator and optimizer")
+flags.mark_flag_as_required("bread")
+flags.DEFINE_integer("batch", 16, "Batch: indicates batch size")
 
 DATA_PATH = '/flour/noCapsnoRepeatsSingleExampleProtos/'
+# DATA_PATH  = '/home/ostap/Documents/fontbakers/data/train/'
```
Comment (ccurro) [inline src/train.py]: Remove. We can make path into flag in subsequent PR with this as defualt

File: src/train.py [hunk 12]
```diff
@@ -101,18 +131,48 @@ def main(argv):
             labels = [CLASS_INDEX[label] for label in labels]
             labels = torch.tensor(labels).to(DEVICE)
 
-            # Zero the parameter gradients
-            optimizer.zero_grad()
+            style_vec = torch.rand((FLAGS.batch, 100), device=DEVICE)
```
Comment (ccurro) [inline src/train.py]: style vect size should be configurable. but leave for now

File: src/train.py [hunk 13]
```diff
@@ -101,18 +131,48 @@ def main(argv):
...
+            char_vec = char_vector(FLAGS.batch, DEVICE)
+
+            # add index for fake character
+            fake_vec = 71 * torch.ones(FLAGS.batch, 1).to(DEVICE)
```
Comment (ccurro) [inline src/train.py]: The Characters array is in this file we shouldnt' have 71's strewn around

Comment (ArianaFreitag) [reply src/train.py]: huh?

Comment (ccurro) [reply src/train.py]: There's an array of the characters we care about. we can do len(CHARACTERS) + 1 rather than having a bunch of hard coded 71s
